CASSJAVA-30: Update CONTRIBUTING.md (#2041)
apache/cassandra-java-driver: CONTRIBUTING.md

under the License.
-->

> # Contributing to the Cassandra Java Driver

Reviewer: ```suggestion
# Contributing to the Apache Cassandra Java Driver
```

Thank you for your interest in contributing!  

## Table of Contents
- [Ways to Contribute](#ways-to-contribute)
- [Contribution Process](#contribution-process)
  - [Reporting Issues](#reporting-issues)
  - [Submitting Changes (Pull Requests)](#submitting-changes-pull-requests)
- [Development Setup](#development-setup)
  - [Prerequisites](#prerequisites)
  - [Building the Project](#building-the-project)
  - [Running Tests](#running-tests)
- [Coding Guidelines](#coding-guidelines)
  - [General](#general)
  - [Code Formatting and License Headers](#code-formatting-and-license-headers)
  - [Javadoc](#javadoc)

>   - [Logging](#logging)

Reviewer: Looks like the actual section is titled `Logs` so this link is broken, but I like `Logging` better so mind changing the section name at line 231?

Reviewer 2: 👍 I very much agree that "Logging" is a better description for that section.

Author: updated

  - [Concurrency annotations](#concurrency-annotations)
  - [Nullability annotations](#nullability-annotations)
- [Coding Guidelines for Tests](#coding-guidelines-for-tests)
  - [Coding style -- test code](#coding-style----test-code)
  - [Unit tests](#unit-tests)
  - [Integration tests](#integration-tests)

## Ways to Contribute

There are many ways to contribute, including:

- **Bug Reports**: Identify incorrect behavior, inconsistencies, or regressions in the driver. Provide reproduction steps when possible.
- **Feature Requests**: Propose improvements or new functionality. Please describe the use case (not just a proposed API).
- **Documentation Improvements**: Enhance guides, examples, javadocs, or configuration explanations.
- **Pull Requests**: Submit fixes, enhancements, performance improvements, or refactorings.
- **Testing Contributions**: Add missing tests, improve coverage, or enhance test infrastructure.
- **Support & Triage**: Help evaluate reported issues or contribute to discussions.
- **Verify Releases**: Verify the release artifacts work correctly in your environment, when a release is proposed in the mailing list.

### Communication

> 1. **Mailing List**: https://groups.google.com/a/lists.datastax.com/g/java-driver-user

Reviewer 2: This Google Group is no longer a means of communication for this project.  The Python and CPP drivers both point users at ASF Slack instead (see [here](https://github.com/apache/cassandra-python-driver/blob/trunk/README.rst?plain=1#L76-L79) as an example) and we should as well.

It looks like the top-level README also [needs to get a similar update](https://github.com/apache/cassandra-java-driver/blob/4.x/README.md?plain=1#L79)... that's outside the scope of this PR and can be handled separately.

> 2. **Slack**: #cassandar-drivers channel in the Apache Software Foundation [Slack](https://infra.apache.org/slack.html). You can ask for an invite to the ASF Slack workspace in the mailing lists.

Reviewer: Small suggestion in spelling and also direct linking to the channel.
```suggestion
2. **Slack**: [#cassandra-drivers](https://the-asf.slack.com/archives/C05LPRVNZV1) channel in the Apache Software Foundation [Slack](https://infra.apache.org/slack.html). You can ask for an invite to the ASF Slack workspace in the mailing lists.
```

3. **GitHub Repository**: https://github.com/apache/cassandra-java-driver
4. **Slack**: #cassandar-drivers channel, if you are in the Apache Software Foundation [Slack](https://infra.apache.org/slack.html)

## Contribution Process

### Reporting Issues

All issues must be tracked in **Apache JIRA**:  
<https://issues.apache.org/jira/projects/CASSJAVA>

When filing an issue:

- Clearly describe the problem, expected behavior, and actual behavior.
- Include driver version, Java version, and Cassandra cluster details.
- Add reproduction steps or a minimal test case if possible.
- Use the appropriate issue type (Bug, Improvement, New Feature, etc.).
- Set the correct components (e.g., `core`, `mapper-runtime`, `quarkus`).

Committers will help refine the ticket if needed.

### Submitting Changes (Pull Requests)

All code changes require:

> 1. **A corresponding JIRA ticket**  

Reviewer 2: We've been asked this question a few times and in general our policy is a bit loose here.  We don't _require_ a JIRA ticket for changes and I'm not sure I want to give the impression that that's a roadblock for users contributing code.  We should certainly say that users are _encouraged_ to create a JIRA ticket describing the problem and perhaps including what they'd like to see done to fix it... and then follow that up with a PR.

   Include the JIRA key in the PR title, e.g.:  
   `CASSJAVA-40: Driver testing against Java 21`
   
2. **A pull request on GitHub**  
   Repository: <https://github.com/apache/cassandra-java-driver>

3. **Tests**  
   Every fix or feature should include or update tests. PRs without tests are rarely accepted.

4. **Documentation updates**  
   Update manual, javadocs, examples, or reference docs when applicable.

5. **Passing CI**  
   PRs must pass all CI jobs unless reviewers explicitly allow exceptions.

**Do not** mix unrelated changes in one PR—keep contributions focused.

**Do not** base a PR on another one.

> **Do not** squash commits before the PR is ready to merge.

Reviewer 2: Out of curiosity... why not?  If a user squashes commits on their PR branch and pushes what's the harm done?

Author: For example, if there is a PR that has one commit for a massive rename, and subsequent commits doing other things, then we want to keep that commit history in case we want to revert the massive rename. 
But this is one of the many things in this CONTRIBUTING.md that I find trivial.

Reviewer: agreed, makes it harder to understand what has changed if the history changes during a review.

Reviewer 2: I guess I don't see this as a huge problem but agree it doesn't hurt anything to prohibit it in the general case.

---

## Development Setup

> ### Prerequisites

Reviewer 2: This raises an interesting point: the pre-reqs mentioned below apply to 4.x but are different for 3.x.  I don't know that we need to enumerate a full set of separate requirements for 3.x distinct from those for 4.x but we probably should at least say (1) these reqs apply to 4.x and (2) 3.x may have different (and likely older) requirements.

Author: If they want a development set up for 3.x then they should just look for the CONTRIBUTING.md in the 3.x branch. https://github.com/apache/cassandra-java-driver/blob/3.x/CONTRIBUTING.md

Reviewer 2: Ah, entirely fair point... we can ignore this one!

- **Java 8+** 

> - **Maven**

Reviewer 2: We have a minimum Maven version that's supported for the project; probably need to explicitly include that here.  CASSJAVA-102 bumped that minimum up to 3.8.1 at least for 4.x.

> - **Java 8+** 

Reviewer: Something that i bumped into is that it is very difficult to run with an up to date JDK 8 as newer OpenJDK build fail on Cassandra setting a very low stack size:

```
ccm start --wait-for-binary-proto --config-dir=/var/folders/ts/m0zplxmx30dcyp3ffzkgwpyw0000gn/T/ccm8054909345564425947 -v
b'\nThe stack size specified is too small, Specify at least 640k\n'
```

I wonder if for this reason, and that newer Cassandra versions (5.0+) require JDK11 we just advise folks use Java 11 at point?

```suggestion
- **Java 8+** (Java 11+ is recommended)
```

Reviewer: Actually, lets keep this at Java 8 for now as I had some issues running with JDK 11, i'll create a follow on JIRA to address this :+1:

### Building the Project

- Ensure Maven is installed and you are using Java 8.
- Build the project with:
  ```
  mvn clean package -DskipTests
  ```
- If using an IDE like IntelliJ and encountering issues with guava-shaded classes:
   - Run:
     ```
     mvn clean install -DskipTests
     ```

>    - If IntelliJ uses a different Maven version, use the Maven window in IntelliJ: under `Lifecycle`, click `clean` and then `install`.

Reviewer 2: IIRC this is a two-step process:

1. The user has to install a local version of the shaded Guava JAR into their Maven cache (which is what the command above does)
2. The user has to ignore the project in the "Maven" flyout window in the upper right (because we want them to fall back into the version that's now installed in their Maven cache).

Author: From my experience, I didn't need to ignore the IntelliJ built-in maven, I just had to make the IntelliJ built-in maven to install the cache, too.

Reviewer 2: Hmmmm, that was very definitely _not_ adequate for me... but I'm fine rolling with this for now.  We can deal with it on a case-by-case basis via Slack and update this if it becomes a problem (which I doubt it will).

### Running Tests

#### Unit Tests

```bash

> mvn clean install -DskipTests

Reviewer: Was curious about this, is running:

```
mvn clean install -DskipTests
```

before 

```
mvn test
```

Actually necessary now?

I just purged my maven repo (`rm -rf ~/.m2/repository`) and ran `mvn test` as is and it worked.

Reviewer: looks like this was required on #2045, but i suspect if on more recent commit this isn't needed anymore.

Reviewer: if i rebase 2045 everything works fine

Reviewer 2: This used to be required because of the need to have the shaded Guava JAR in place when compiling tests IIRC... I think I've still seen that behaviour fairly recently.

mvn test
```

#### Integration Tests

1. Install Cassandra Cluster Manager (CCM) following its [README](https://github.com/apache/cassandra-ccm).

> 2. On MacOS only, enable loopback aliases:

Reviewer 2: This is added here for... Simulacron testing, right?  Might be worth calling that out explicitly.  As it stands now this text appears to refer to something required for ccm testing which I don't think is correct.

Author: Isn't loopback aliases for CCM testing? I always need to turn on loopback aliases whenever I use CCM

Reviewer: yeah its needed for simulacron too since it uses the same local IP addresses.  You don't have to install anything though, i guess we can note it here:

```suggestion
2. **MacOS only**, for CCM and Simulacron-based tests, enable loopback aliases:
```

Reviewer 2: I don't have personal knowledge of the requirement; I don't run any of this on OS X.  I'm just going by what the [original CONTRIBUTING doc said](https://github.com/apache/cassandra-java-driver/blob/4.19.2/CONTRIBUTING.md?plain=1#L269-L275) on the topic.

Author: updated

```shell
for i in {2..255}; do sudo ifconfig lo0 alias 127.0.0.$i up; done
   ```

>    Note: This may slow down networking. To remove the aliases after testing:

Reviewer: Unfortunately this looks to have gotten even worse on newer mac os versions. I wonder if we can get away with creating less aliases...will see if I can find a minimum.

Reviewer: no need to change this for now, will find the minimum and can address separately

   ```shell
   for i in {2..255}; do sudo ifconfig lo0 -alias 127.0.0.$i up; done
   ```
3. Run integration tests:
   ```
   mvn clean verify
   ```
   To target a specific Cassandra version or distribution:
   ```

>    mvn verify -Dccm.version=3.11.0

Reviewer: We should bump this to a newer version, primarily because a version this old won't work on arm based Macs, and also its nice to test against latest (more tests will run)

```suggestion
   mvn verify -Dccm.version=5.0.6
```

However, I think for this to work we need a pointer to JDK 11, so maybe for now we target 4.1.11, and can think about how to update docs around running tests against newer cassandra versions as well which require newer JDKs:

```suggestion
   mvn verify -Dccm.version=4.1.10
```

Reviewer: After some testing , I think we can just use 5.0.6 with JDK 11 just fine, we should just advise using Java 11 in the docs (see above), i think it will be the most dependable and also has the advantage of being able to run most of the tests.

Reviewer: Lets keep this @ 3.11.0 for now, as running with JDK 11 doesn't work reliably, will address in follow on

Reviewer 2: It's meant as an example here for the user; presumably they'd tweak it for their own use case when they actually run it.  There's certainly no _harm_ in bumping this up to a newer version but even using 3.11.0 seems to illustrate the point well enough.

- [Development Environment Setup](#development-environment-setup)
- [To Submit a Pull Request](#to-submit-a-pull-request)
- [Coding Guide](#coding-guide)

## Bugs, Features, and Questions

- To report bugs or request features, please file a ticket on our [JIRA](https://issues.apache.org/jira/projects/CASSJAVA).
- For questions, use our [user mailing list](https://groups.google.com/a/lists.datastax.com/g/java-driver-user).

## Development Environment Setup

### Build and IDE Configuration

- Ensure Maven is installed and you are using Java 8.
- Build the project with:
  ```
  mvn clean package -DskipTests
  ```
- If using an IDE like IntelliJ and encountering issues with guava-shaded classes:
    - Run:
      ```
      mvn clean install -DskipTests
      ```
    - If IntelliJ uses a different Maven version, use the Maven window in IntelliJ: under `Lifecycle`, click `clean` and then `install`.

### Running the Tests

#### Unit Tests

- Ensure you are using Java 8.
- Run:
  ```
  mvn clean install -DskipTests
  mvn test
  ```

#### Integration Tests

1. Install Cassandra Cluster Manager (CCM) following its [README](https://github.com/apache/cassandra-ccm).

> 2. On macOS, enable loopback aliases:

Reviewer 3: Maybe explicitly mention `On MacOS only, ...`

   ```
   mvn verify -Dccm.version=3.11.0
   ```

### Code Formatting and License Headers

- We follow the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html). See [google-java-format](https://github.com/google/google-java-format) for IDE plugins.
- To format code:
  ```
  # Java files
  mvn fmt:format
  # XML files
  mvn xml-format:xml-format
  # License headers
  mvn license:format
  ```

### Pre-commit Hook (Highly Recommended)

- Make `pre-commit.sh` executable:
  ```
  chmod +x pre-commit.sh
  ```
- Set up the pre-commit hook:
  ```
  ln -s ../../pre-commit.sh .git/hooks/pre-commit
  ```
- This script will format files and run unit tests before each commit.

## To Submit a Pull Request

- Pull request titles must follow the format: `{JIRA-ticket-number}: {JIRA-ticket-title}`  
  Example: `CASSJAVA-40: Driver testing against Java 21`
- Before merging, squash your commits into a single commit with a message formatted as:
  ```
  {JIRA-ticket-number}: {JIRA-ticket-title}
  patch by {your-name}; reviewed by {approver-1} and {approver-2} for {JIRA-ticket-number}
  ```

## Coding Guide
### Javadoc

- **API packages:** must be documented.

> - **Internal packages:** optional but encouraged—add a short class-level comment explaining the component’s role.

Reviewer 3: `encouraged  - add `

You don’t need to document every method or parameter. Skip obvious stuff like:

```
/**
 * @return the name
 */
String getName();
```
### Logs
We use **SLF4J**:

```
private static final Logger LOG = LoggerFactory.getLogger(TheEnclosingClass.class);
```

Logs help two groups:
1. **Ops** – running the app in production.
2. **Developers** – debugging issues.

**Log levels:**

- **ERROR** – critical, needs action, not the ones re-thrown to the client
- **WARN** – recoverable but should be fixed.
- **INFO** – normal but useful events.
- **DEBUG** – detailed activity (e.g., state changes).
- **TRACE** – per-request details.

To change the log level when running tests, modify the `src/test/resources/logback-test.xml`.

**Log prefix** shows origin, e.g.:  
`[s0|90232530|0]` (session name | hash code of the CqlRequestHandler instance | number of request attempts)

### Integration Tests
Testing against an external process, using either of the tools:
- [Simulacron](https://github.com/datastax/simulacron): simulates Cassandra nodes on loopback. You “prime” it with expected query results.  
  *Example*: `NodeTargetingIT`.

> - [CCM](https://github.com/pcmanus/ccm): launches actual Cassandra nodes locally. The `ccm` executable must be in the path. Pass variables like `-Dccm.version=5.0.0 -Dccm.dse=false` to specify the Cassandra version or distribution.

Reviewer 3: Should be `-ccm.distribution=dse`, not `-Dccm.dse=false`.

Apply to the whole class and are reused between methods. Preferred for efficiency.

```java
private static final CcmRule CCM_RULE = CcmRule.getInstance(); 
private static final SessionRule<CqlSession> SESSION_RULE = SessionRule.builder(CCM_RULE).build();  

@ClassRule 
public static final TestRule CHAIN = RuleChain.outerRule(CCM_RULE).around(SESSION_RULE);
```

**Method-level rules** (`@Rule`):  
Re-initialize for each test method. Use for method-specific requirements or clean states.

```java
private CcmRule ccmRule = CcmRule.getInstance(); 
private SessionRule<CqlSession> sessionRule = SessionRule.builder(ccmRule).build();  

@Rule 
public TestRule chain = RuleChain.outerRule(ccmRule).around(sessionRule);
```
### Miscellaneous
1. Nullability: Please use the [Spotbugs annotations](https://spotbugs.github.io) to document nullability of new class or interface using the `@Nullable` and `@NonNull` annotations from the package `edu.umd.cs.findbugs.annotations`.
2. Concurrency annotations: Please use the [JCIP annotations](http://jcip.net/annotations/doc/index.html) to document thread-safety.
3. Static imports: Please only use static imports for AssertJ, Mockito, and Awaitility.

> 4. `toString()`: only for debug in development. Use `toCqlLiteral()` is applicable for logs.

Reviewer 3: Shorten format does not tell the meaning of previous long description. Not all classes would have `toCqlLiteral()`.

## License headers

The build will fail if some license headers are missing. To update all files from the command line,
run:

```
mvn license:format
```

## Pre-commit hook (highly recommended)
 
Ensure `pre-commit.sh` is executable, then run:

```
ln -s ../../pre-commit.sh .git/hooks/pre-commit
```

This will only allow commits if the tests pass. It is also a good reminder to keep the test suite
short.

Note: the tests run on the current state of the working directory. I tried to add a `git stash` in
the script to only test what's actually being committed, but I couldn't get it to run reliably
(it's still in there but commented). Keep this in mind when you commit, and don't forget to re-add
the changes if the first attempt failed and you fixed the tests.

## Speeding up the build for local tests

If you need to install something in your local repository quickly, you can use the `fast` profile to
skip all "non-essential" checks (licenses, formatting, tests, etc):

```
mvn clean install -Pfast
```

You can speed things up even more by targeting specific modules with the `-pl` option:

```
mvn clean install -Pfast -pl core,query-builder,mapper-runtime,mapper-processor,bom
```

> Please run the normal build at least once before you push your changes.

Reviewer: Noticed this section was removed, but we did update pre-commit.sh should we keep this doc?

of call sites can be reviewed in a few seconds, but if it's part of a larger diff it gets mixed up
with more complex changes (that might affect the same lines), and reviewers have to check every
line.

Commit message subjects start with a capital letter, use the imperative form and do **not** end
with a period:

* correct: "Add test for CQL request handler"
* incorrect: "~~Added test for CQL request handler~~"
* incorrect: "~~New test for CQL request handler~~"

Avoid catch-all messages like "Minor cleanup", "Various fixes", etc. They don't provide any useful
information to reviewers, and might be a sign that your commit contains unrelated changes.
 
We don't enforce a particular subject line length limit, but try to keep it short.

You can add more details after the subject line, separated by a blank line. The following pattern
(inspired by [Netty](http://netty.io/wiki/writing-a-commit-message.html)) is not mandatory, but
welcome for complex changes:

```
One line description of your change
 
Motivation:

Explain here the context, and why you're making that change.
What is the problem you're trying to solve.
 
Modifications:

Describe the modifications you've done.
 
Result:

> After your change, what will change.

Reviewer: I think this was generally good advice, but not something that has really been enforced for a while, so +1 to removing.

```

## Pull requests

Like commits, pull requests should be focused on a single, clearly stated goal.

Don't base a pull request onto another one, it's too complicated to follow two branches that evolve
at the same time. If a ticket depends on another, wait for the first one to be merged. 

If you have to address feedback, avoid rewriting the history (e.g. squashing or amending commits):
this makes the reviewers' job harder, because they have to re-read the full diff and figure out
where your new changes are. Instead, push a new commit on top of the existing history; it will be
squashed later when the PR gets merged. If the history is complex, it's a good idea to indicate in

> the message where the changes should be squashed:

Reviewer 2: Interesting that we used to have this instruction in the CONTRIBUTING doc.  It obviously doesn't apply anymore now that we've moved away from squash merges... but it _does_ suggest to me that this might be a good place to provide instructions to a user about how do a squash via an interactive rebase.  We _can_ do that for user-contributed PRs but I'd kind of prefer to put that on the contributors if we can.
